### PR TITLE
GODRIVER-1576 fix non-addr type with pointer implementation for bson

### PR DESCRIPTION
**Pull request by iwysiu**
Right now this doesn't store CondAddrEncoders in the typeEncoders map. Would it make sense to store half-filled CondAddrEncoders?

### CODE REVIEW

File: bson/bsoncodec/cond_addr_codec.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+)
+
+// condAddrEncoder is the encoder used when a pointer to the encoding value has an encoder.
+type condAddrEncoder struct {
+	canAddrEnc ValueEncoder
+	elseEnc    ValueEncoder
+}
+
+var _ ValueEncoder = &condAddrEncoder{}
```
Comment (divjotarora) [inline bson/bsoncodec/cond_addr_codec.go:21]: nit: For pointer types, we usually format the type assertions as `var _ ValueEncoder = (*condAddrEncoder)(nil)`. Same for the `ValueDecoder` assertion below.

File: bson/bsoncodec/registry.go [hunk 2]
```diff
@@ -359,14 +375,21 @@ func (r *Registry) lookupTypeEncoder(t reflect.Type) (ValueEncoder, bool) {
...
 	if t == nil {
 		return nil, false
 	}
 	for _, ienc := range r.interfaceEncoders {
-		if t.Implements(ienc.i) || reflect.PtrTo(t).Implements(ienc.i) {
+		if t.Implements(ienc.i) {
 			return ienc.ve, true
 		}
+		if allowAddr && t.Kind() != reflect.Ptr && reflect.PtrTo(t).Implements(ienc.i) {
+			defaultEnc, found := r.lookupInterfaceEncoder(t, false)
```
Comment (divjotarora) [inline bson/bsoncodec/registry.go:387]: Is this nested call to `lookupInterfaceEncoder` needed for the case where `*T` implements this the interface we're checking for but `T` implements some other interface that's later in the further ahead in the `interfaceEncoders` list?

File: bson/bsoncodec/registry.go [hunk 3]
```diff
@@ -424,13 +463,14 @@ func (r *Registry) lookupTypeDecoder(t reflect.Type) (ValueDecoder, bool) {
...
 
-func (r *Registry) lookupInterfaceDecoder(t reflect.Type) (ValueDecoder, bool) {
+func (r *Registry) lookupInterfaceDecoder(t reflect.Type, allowAddr bool) (ValueDecoder, bool) {
 	for _, idec := range r.interfaceDecoders {
-		if !t.Implements(idec.i) && !reflect.PtrTo(t).Implements(idec.i) {
-			continue
+		if t.Implements(idec.i) {
+			return idec.vd, true
+		}
+		if allowAddr && t.Kind() != reflect.Ptr && reflect.PtrTo(t).Implements(idec.i) {
```
Comment (divjotarora) [inline bson/bsoncodec/registry.go:471]: Why isn't a recursive call to `lookupInterfaceDecoder(t, false)` needed here as it is in `lookupInterfaceEncoder`?

File: bson/bsoncodec/registry.go [hunk 4]
```diff
@@ -322,14 +322,30 @@ func (r *Registry) LookupEncoder(t reflect.Type) (ValueEncoder, error) {
 		if enc == nil {
 			return nil, ErrNoEncoder{Type: t}
 		}
-		return enc, nil
+		// if dec is a condAddrEncoder, we still need to fill in elseEnc
```
Comment (divjotarora) [inline bson/bsoncodec/registry.go:325]: I'm not sure I understand this. Let's say I have type `T` and `*T` implements `Marshaler`. The first time I try to marshal value of type `T`, this should call `lookupInterfaceEncoder`, which will return:
```
condAddrEncoder{
    canAddrEnc: MarshalerEncodeValue
    elseEnc: StructCodec.EncodeValue
}
```
and we'll store that in `typeEncoders` as a cached entry. The next time I try to marshal something of type `T`, why do we need to go through that whole process again? Are there cases where the cached `condAddrEncoder` is not correct?

File: bson/bsoncodec/cond_addr_codec.go (hunk 1 above)
Comment (iwysiu) [reply bson/bsoncodec/cond_addr_codec.go]: done

File: bson/bsoncodec/registry.go (hunk 2 above)
Comment (divjotarora) [reply bson/bsoncodec/registry.go:371]: I'm fine with this. Let's add a comment to explain it, though.

File: bson/bsoncodec/registry.go (hunk 3 above)
Comment (divjotarora) [reply bson/bsoncodec/registry.go:439]: Same thing here about leaving a comment. Also, I'm not sure if the same logic is needed for the decode code path. The top-level `Unmarshal` functions only allow a pointer or map to be specified. Does `encoding/json` have a conditional decoder type?

Comment (iwysiu) [reply bson/bsoncodec/registry.go:441]: encoding/json has a TextUnmarshaler too

File: bson/bsoncodec/cond_addr_codec_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+func TestCondAddrCodec(t *testing.T) {
+	var inner int
+	canAddrVal := reflect.ValueOf(&inner)
+	addressable := canAddrVal.Elem()
+	unaddressable := reflect.ValueOf(inner)
+	rw := &bsonrwtest.ValueReaderWriter{}
+
+	t.Run("Encode", func(t *testing.T) {
+		invoked := 0
+		encode1 := func(EncodeContext, bsonrw.ValueWriter, reflect.Value) error {
```
Comment (divjotarora) [inline bson/bsoncodec/cond_addr_codec_test.go:27]: nit: give these more meaningful names like `addressEncode` and `valueEncode` or add some comments.

File: bson/bsoncodec/cond_addr_codec_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+			val     reflect.Value
+			invoked int
+		}{
+			{"canAddr", addressable, 1},
+			{"else", unaddressable, 2},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				err := condEncoder.EncodeValue(EncodeContext{}, rw, tc.val)
+				assert.Nil(t, err, "StringCodec.EncodeValue error: %v", err)
```
Comment (divjotarora) [inline bson/bsoncodec/cond_addr_codec_test.go:48]: nit: This shouldn't say `StringCodec`. Copy/paste error maybe?

File: bson/bsoncodec/cond_addr_codec_test.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		t.Run("error", func(t *testing.T) {
+			errEncoder := newCondAddrEncoder(ValueEncoderFunc(encode1), nil)
+			err := errEncoder.EncodeValue(EncodeContext{}, rw, unaddressable)
+			want := ErrNoEncoder{Type: unaddressable.Type()}
+			assert.Equal(t, err, want, "expected error %v, got %v", want, err)
+		})
+	})
+	t.Run("Decode", func(t *testing.T) {
+		invoked := 0
+		decode1 := func(dctx DecodeContext, vr bsonrw.ValueReader, val reflect.Value) error {
```
Comment (divjotarora) [inline bson/bsoncodec/cond_addr_codec_test.go:63]: Same thing here about meaningful function names or comments.

File: bson/bsoncodec/cond_addr_codec_test.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			val     reflect.Value
+			invoked int
+		}{
+			{"canAddr", addressable, 1},
+			{"else", unaddressable, 2},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
+				err := condDecoder.DecodeValue(DecodeContext{}, rw, tc.val)
+				assert.Nil(t, err, "StringCodec.DecodeValue error: %v", err)
```
Comment (divjotarora) [inline bson/bsoncodec/cond_addr_codec_test.go:84]: Same thing here about `StringCodec`.

File: bson/bsoncodec/cond_addr_codec_test.go (hunk 7 above)
Comment (iwysiu) [reply bson/bsoncodec/cond_addr_codec_test.go:63]: done

File: bson/bsoncodec/cond_addr_codec_test.go (hunk 5 above)
Comment (iwysiu) [reply bson/bsoncodec/cond_addr_codec_test.go:27]: done

File: bson/bsoncodec/cond_addr_codec_test.go (hunk 6 above)
Comment (iwysiu) [reply bson/bsoncodec/cond_addr_codec_test.go]: done

File: bson/bsoncodec/cond_addr_codec_test.go (hunk 8 above)
Comment (iwysiu) [reply bson/bsoncodec/cond_addr_codec_test.go]: done

**Comment by divjotarora** [comment]
LGTM stands @iwysiu

**Pull request merged**